astropy/coordinates/builtin_frames/cirs_observed_transforms.py

```diff
@@ -22,11 +22,13 @@
 
 @frame_transform_graph.transform(FunctionTransformWithFiniteDifference, CIRS, AltAz)
 def cirs_to_altaz(cirs_coo, altaz_frame):
-    if np.any(cirs_coo.obstime != altaz_frame.obstime):
-        # the only frame attribute for the current CIRS is the obstime, but this
-        # would need to be updated if a future change allowed specifying an
-        # Earth location algorithm or something
-        cirs_coo = cirs_coo.transform_to(CIRS(obstime=altaz_frame.obstime))
+
+    obsgeoloc, obsgeovel = altaz_frame.location.get_gcrs_posvel(altaz_frame.obstime)
+    if (np.any(cirs_coo.obstime != altaz_frame.obstime) or
+            np.any(cirs_coo.obsgeoloc != obsgeoloc):
+        cirs_coo = cirs_coo.transform_to(CIRS(obstime=altaz_frame.obstime,
```

Comment:
One way to speed up a straight through CIRS->AltAz transformation would be to allow the AltAz `obstime` and `location` to be `None` which would then just use the CIRS frame attributes.

Reply:
Sorry @mkbrewer - this isn't a very useful suggestion since it requires re-writing all of how astropy works from the ground up. By construction you get an AltAz coordinate at the location and time specified on the AltAz frame.

Reply:
Isn't that because the CIRS frame didn't have a `location` attribute before?

Reply:
In the current framework, we can only define an `AltAz` frame using explicit locations, i.e.,  `AltAz(obsgeoloc=..., obsgeovel=...)`, but there is no reason the code couldn't realize when doing the transform that those are identical to the CIRS ones. 

Maybe more directly similar, I think at least at the `SkyCoord.transform_to` level, we could have at least the string version ('altaz') mean "create an `AltAz` frame with attributes as available on the `SkyCoord`" (at the `BaseFrame.transform_to` level, I think we want to stick to being super explicit). So, I think this could be made to work quite conveniently!

Anyway, would be for a follow-up PR!

Reply:
It works that way for all transforms. The target frame sets the properties.  What you suggest would change the API, and we don’t do that

Reply:
I see. So it's a matter of philosophy then. I've always wondered why I needed to specify the time and location again when I wanted AltAz from the result of a call to `get_body()`.

Reply:
I think we should distinguish frames from `SkyCoord` here. In frames, everything is explicit, while `SkyCoord` can carry attributes not directly relevant to a given frame (say, `obsgeoloc` for ICRS). I do think it is very possible for the `SkyCoord` output of `get_body()` to pass on the time and location in `.transform_to` - though obviously we actually need to implement it!

Reply:
I should think that it would be generally advisable to allow data to flow downstream unimpeded rather than encountering data at the end of a chain that requires the entire chain to be recomputed. I don't really see the application of computing coordinates at one place and time only to change that place and time further downstream, but I must be missing something there.

Reply:
It is really mostly about where data is stored. The `ICRS` class has only coordinates, nothing about an observer, so it cannot transform to, say, `AltAz`, without adding those pieces of information, but it can transform to `AltAz(obstime=..., ...)`.

On the other hand,  `SkyCoord`, can hold extra information.

Of course, really this should all be an implementation detail - `SkyCoord` is the user-facing class. And I think it is actually meant to work on `SkyCoord`, but doesn't - see #10996.